presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -788,14 +800,45 @@ private synchronized void cacheLastDataCommitTimes(Map<String, Optional<Partitio
         }
     }
 
-    public synchronized void setPartitionLeases(MetastoreContext metastoreContext, String databaseName, String tableName, Map<String, String> partitionNameToLocation, Duration leaseDuration)
+    /**
+     * Set partition leases for a given table in the metastore.
+     *
+     * @param metastoreContext The context of the metastore
+     * @param databaseName The name of the database the table is in
+     * @param tableName The name of the table
+     * @param partitionNameToLocation A map of partition names to their location
+     * @param leaseDuration The duration of the lease
+     * @param leaseRenewalPeriod The period at which the lease should be renewed
+     */
+    public synchronized void setPartitionLeases(
+            MetastoreContext metastoreContext,
+            String databaseName,
+            String tableName,
+            Map<String, String> partitionNameToLocation,
+            Duration leaseDuration,
+            Duration leaseRenewalPeriod)
     {
         checkReadable();
         Table table = getTable(metastoreContext, databaseName, tableName)
                 .orElseThrow(() -> new TableNotFoundException(new SchemaTableName(databaseName, tableName)));
         boolean isPartitioned = table.getPartitionColumns() != null && !table.getPartitionColumns().isEmpty();
         if (table.getTableType().equals(MANAGED_TABLE) && isPartitioned && leaseDuration.toMillis() > 0) {
             delegate.setPartitionLeases(metastoreContext, databaseName, tableName, partitionNameToLocation, leaseDuration);
+            log.info("Setting partition lease for database %s, table %s, duration %s", databaseName, tableName, leaseDuration);
+
+            // If a lease renewal period is specified, schedule a task to periodically renew the lease
+            if (leaseRenewalPeriod.toMillis() > 0) {
+                Runnable runnable = () -> {
+                    log.info("Extending partition lease for database %s, table %s by %s", databaseName, tableName, leaseDuration);
+                    try {
+                        delegate.setPartitionLeases(metastoreContext, databaseName, tableName, partitionNameToLocation, leaseDuration);
+                    }
+                    catch (Exception e) {
+                        log.warn(e, "Unable to extend partition lease for database %s, table %s", databaseName, tableName);
+                    }
+                };
+                leaseRenewalExecutor.get().scheduleAtFixedRate(runnable, leaseRenewalPeriod.toMillis(), leaseRenewalPeriod.toMillis(), TimeUnit.MILLISECONDS);
```

Comment:
so when will this runnable be stopped from running again and again ? when the query completes ?

Reply:
This is a great point. I approached the problem from the perspective of Presto on Spark, where this feature would be extremely useful. But the execution model is different there as it's a single JVM per each query and when the query is finished the process is killed. I need to figure out how to handle that for Presto multi-tenant clusters. Feel free to reach out if you have any suggestions.